### ISSUE TITLE
eeschema: annotate hierarchy: no control over sheet numbers

### ISSUE THREAD

**Issue by marekr** [marekr/kicad-issue-test#207]
When I place several hierarchial sheets, I can't alter (not even see) the sheetnumber. When i then annotate, the resulting component references are not always what I wanted.

Suppose I have 2 functional blocks, let's call them A.sch and B.sch. I would like to create a PCB with 3A's, and 2 B's. Here's what happens when I create them in this order:
1-create the main sheet hierarchy.sch
2-place hierarchial sheet A.sch (called A1)
3-place hierarchial sheet B.sch (called B1)
4-copy hierarchial sheet A1 and rename to A2
5-copy hierarchial sheet A1 and rename to A3
6-copy hierarchial sheet B1 and rename to B2

When I now annotate (entire schematic, start to sheet number *100) then I get the following result:
A1: components start with 201
A2: components start with 401
A3: components start with 501
B1: components start with 301
B2: components start with 601
Topmost sheet: components start with 101 

Is it possible to make the sheet number editable, so I can take the above schematic, and change the numbers so they become like this?
A1: components start with 101
A2: components start with 201
A3: components start with 301
B1: components start with 401
B2: components start with 501
Topmost sheet: components start with 601

Best regards,
Cedric

**Comment by marekr** [comment]
(by cedric-dewijs-telfort)
Sorry, I forgot to post my version:
Application: KiCad
Version: (2013-03-04 BZR 3984)-stable
Build: wxWidgets 2.9.4 (wchar_t,compiler with C++ ABI 1002,GCC 4.7.2,wx containers,compatible with 2.8)
Platform: Windows XP (build 2600, Service Pack 3), 32 bit, Little endian, wxMSW
Boost version: 1.53.0
Options: USE_PCBNEW_NANOMETRES=ON
         KICAD_GOST=OFF
         USE_WX_GRAPHICS_CONTEXT=OFF
         USE_WX_OVERLAY=OFF
         KICAD_SCRIPTING=OFF
         KICAD_SCRIPTING_MODULES=OFF
         KICAD_SCRIPTING_WXPYTHON=OFF

**Comment by marekr** [comment]
(by zeilhofer)
I have the same problem. 

At lest I found a workaround:
Edit the .sch file in a text editor of your choice, and rearange the blocks of $Sheet to $EndSheet in your favour order. 

When you reopen the file in eeshema, the sheet order has changed. 

It would be a necessary feature for eeschema. 
My problem especially was, that I changed a large project, and added some sheets. And then the sheet numbers got mixed up, so that for example all the U6xx components where on sheet 3. And the newly added parts on this sheet got the name U3xx, and didn't match to their neighboured U6xx-parts on the sheet. 

What you can not change with this workaround is the order of sub-sub-sheets. these will allways get a number after the parent sheet (= sub-sheet). When adding sub-sub-sheets to an existing project, this will lead to the problem described above, unless you add such a sub-sub-sheet only to the last sub-sheet. 

could this feature be expected in a near future version?

greets, Karl

**Comment by marekr** [comment]
(by nikbowdler)
This is a problem for me too. 

Being able to manually set the desired page numbering would be a good feature. This could be a parameter in the create/edit sheet box.

**Comment by marekr** [comment]
(by stambaughw)
As long as it can be done with any changes to the current schematic file format then I'm OK with adding this feature.  It will have to be done before the upcoming feature freeze (very soon hopefully) or be put off until the next development cycle.  If it requires changes to the file format to save the ordering information then it will have to wait until the next development cycle because we are planning on changing the schematic file format to an s-expression format like the board file format.  No changes will be allowed to the current file format.

**Comment by marekr** [comment]
(by jp-charras)
I do not think this is possible to do that just by changing the order of sheets (but I can be wrong).

I am seeing at least 2 critical cases:
1 - complex hierarchies
2 - sheets inside sheets.

In every case, you need to have a hierarchy manager, and store the hierarchy structure (order of sheet paths) inside the root schematic file.

**Comment by marekr** [comment]
(by nikbowdler)
No change to the file format would be required, just the algorithm that does the sheet numbering: to use the user input box rather than save order. The box could default to the next available sheet number when creating.

**Comment by marekr** [comment]
(by zeilhofer)
since the KiCad developers are too busy to implement this feature, I have written a simple external tool. 
I'm sorry for not implementing it directly in the KiCad source code, but currently I have not the time to dig into this huge project code. 

Please have a look here:
http://www.zeilhofer.co.at/kicad-sheet-rearranger

Greets, 
Karl

**Comment by marekr** [comment]
(by a-wolfie)
This appears to be a duplicate of earlier reported bug #864122.  But this one has been elevated to wishlist, while the older one is still undecided.
